File: lessons/gen_scripts/ruby/scripts/i20.subroutine.rb
#!/usr/bin/env ruby
# declare pond as an ordinary (local) variable - a method's own locals are
#  isolated by default, so fish's "pond" below can never see or touch this one.
pond      = 500 # pond contains some available fish
$captured = 0   # must stay global; a method can't mutate an outer local

# create method (subroutine) - this "pond" is a brand-new local, unrelated
#  to the one above, no matter how many times fish is called
def fish
  pond       = 500 # local to this method only
  pond      -= 150 # subtract from the local pond (no effect outside)
  $captured += 150 # add to the fish captured
end

# output initial amount of fish in shared resource
puts "We have #{pond} in this pond."

fish # get some fish
puts "Fishing from a local pond... We now have #{pond} in the main pond."

fish # get some fish
puts "Fishing from a local pond... We now have #{pond} in the main pond."

fish # get some fish
puts "Fishing from a local pond... We now have #{pond} in the main pond."

# output result of fish captured from local resource
puts "We now have a total of #{$captured} fish captured"

File: lessons/gen_scripts/perl/scripts/i20.subroutine.pl
#!/usr/bin/env -S perl -w
# declare package globals (no "my", so visible to every sub in this file)
our $pond     = 500; # pond contains some available fish
our $captured = 0;   # captured represents fish captured

# create subroutine
sub fish {
  my $pond   = 500; # "my" makes this a new lexical that shadows the global $pond, local to this sub only
  $pond     -= 150; # subtract from the local pond; the global $pond is untouched
  $captured += 150; # add to the fish captured (still the global)
}

# output initial amount of fish in shared resource
print "We have $pond in this pond.\n";

fish();                                                            # get some fish
print "Fishing from a local pond... We now have $pond in the main pond.\n";

fish();                                                            # get some fish
print "Fishing from a local pond... We now have $pond in the main pond.\n";

fish();                                                            # get some fish
print "Fishing from a local pond... We now have $pond in the main pond.\n";

# output result of fish captured from local resource
print "We now have a total of $captured fish captured\n";
